### PR TITLE
[Auto Parallel] Compatible new comm library upgrade

### PR DESCRIPTION
**Pull request by GhostScreaming**
### PR types
Others

### PR changes
Others

### Description
Pcard-70448

Upgrade fluid operators to be compatible with new communication library. Old communication library is used defaultly. Switch to new communication library by export `FLAGS_dynamic_static_unified_comm=1`. We can use new communication library in default as soon as all communication operators are upgraded compatibly.

### CODE REVIEW

File: python/paddle/distributed/collective.py [hunk 1]
```diff
@@ -319,18 +320,63 @@ def is_available():
...
+            cls._instance = core.TCPStore(
+                master_addr,
+                master_port,
+                is_master,
+                world_size,
+            )
+        return cls._instance
+
+    def __init__(self):
+        pass
```
Comment (LiYuRio) [inline python/paddle/distributed/collective.py]: 这里可以直接用c++的单例吧，没必要自己写了

File: python/paddle/distributed/collective.py [hunk 2]
```diff
@@ -319,18 +320,63 @@ def is_available():
...
+            master_endpoint is not None
+        ), "Please set PADDLE_MASTER enviroment variable."
+    if master_endpoint:
+        master_addr = master_endpoint.split(":")[0]
+        master_port = int(master_endpoint.split(":")[1])
+        global_env = _get_global_env()
+        rank = global_env.rank
+        world_size = global_env.world_size
+        dev_id = global_env.device_id
+        is_master = rank == 0
```
Comment (hitywt) [inline python/paddle/distributed/collective.py]: L323-L336好像是不需要了，在StaticTCPStore里面了

**Review by LiYuRio — DISMISSED** [review]

**Review by hitywt — APPROVED** [review]
LGTM

File: paddle/phi/core/distributed/comm_context_manager.cc [hunk 3]
```diff
@@ -67,7 +77,29 @@ void CommContextManager::CreateNCCLCommContext(
...
+    dev_ctx->PartialInitWithAllocator();
+    auto compute_event =
+        phi::memory_utils::GetCudaEvent(CommContextManager::device_id);
+    auto comm_event =
+        phi::memory_utils::GetCudaEvent(CommContextManager::device_id);
+
+    nccl_comm_context->SetDevContext(std::move(dev_ctx));
+    nccl_comm_context->SetComputeEvent(std::move(compute_event));
+    nccl_comm_context->SetCommEvent(std::move(comm_event));
+  } else {
```
Comment (zhiqiu) [inline paddle/phi/core/distributed/comm_context_manager.cc]: why empty else branch?

File: paddle/phi/core/distributed/comm_context_manager.cc [hunk 4]
```diff
@@ -67,7 +77,29 @@ void CommContextManager::CreateNCCLCommContext(
 
   auto nccl_comm_context =
       std::make_unique<NCCLCommContext>(rank, size, nccl_id);
-  auto& comm_context_manager = CommContextManager::GetInstance();
+
+  if (CommContextManager::device_id != -1) {
```
Comment (zhiqiu) [inline paddle/phi/core/distributed/comm_context_manager.cc:81]: why not use dev_ctx from global dev_ctx pool?

Comment (GhostScreaming) [reply paddle/phi/core/distributed/comm_context_manager.cc:81]: `NCCLCommContext` holds `std::unique_ptr<phi::GPUContext> dev_ctx`, so we create a new object out of `DeviceContextPool` control.

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by LiYuRio — APPROVED** [review]

**Pull request merged**